src/meta/src/stream/stream_graph.rs

```diff
@@ -624,37 +498,78 @@ impl StreamGraphBuilder {
+
+    /// When converting fragment graph to actor graph, we need to know which actors belong to a
     /// fragment.
-    fragment_actors: HashMap<GlobalFragmentId, Vec<LocalActorId>>,
-    /// local actor id
-    next_local_actor_id: u32,
+    fragment_actors: HashMap<GlobalFragmentId, Vec<GlobalActorId>>,
+
+    /// The next local actor id to use.
+    next_local_id: u32,
+
+    /// The global actor id generator.
+    actor_id_gen: GlobalActorIdGen,
 }
 
 impl BuildActorGraphState {
-    fn gen_actor_ids(&mut self, parallel_degree: u32) -> Range<u32> {
-        let start_actor_id = self.next_local_actor_id;
-        self.next_local_actor_id += parallel_degree;
-        start_actor_id..start_actor_id + parallel_degree
+    /// Create an empty state with the given id generator.
+    fn new(actor_id_gen: GlobalActorIdGen) -> Self {
+        Self {
+            stream_graph_builder: Default::default(),
+            fragment_actors: Default::default(),
+            next_local_id: 0,
+            actor_id_gen,
+        }
+    }
+
+    /// Get the next global actor id.
+    fn next_actor_id(&mut self) -> GlobalActorId {
+        let local_id = self.next_local_id;
+        self.next_local_id += 1;
+
+        self.actor_id_gen.to_global_id(local_id)
+    }
+
+    /// Finish the build and return the inner stream graph builder.
+    fn finish(self) -> StreamGraphBuilder {
+        assert_eq!(self.actor_id_gen.len, self.next_local_id);
+        self.stream_graph_builder
     }
 }
 
 /// [`ActorGraphBuilder`] generates the proto for interconnected actors for a streaming pipeline.
 pub struct ActorGraphBuilder {
-    /// Default parallelism.
-    default_parallelism: u32,
+    /// The pre-scheduled distribution for each fragment.
+    // TODO: this is fake now and only the parallelism is used, we should also use the physical
+    // distribution itself after scheduler refactoring.
+    distributions: HashMap<GlobalFragmentId, Distribution>,
 
+    /// The stream fragment graph.
     fragment_graph: StreamFragmentGraph,
 }
 
 impl ActorGraphBuilder {
-    pub fn new(fragment_graph: StreamFragmentGraph, default_parallelism: u32) -> Self {
+    /// Create a new actor graph builder with the given "complete" graph.
+    pub fn new(complete_graph: CompleteStreamFragmentGraph, default_parallelism: u32) -> Self {
+        // TODO: use the real parallel units to generate real distribution.
+        let fake_parallel_units = (0..default_parallelism).map(|id| ParallelUnit {
+            id,
+            worker_node_id: 0,
+        });
+        let distributions =
+            schedule::Scheduler::new(fake_parallel_units, default_parallelism as usize)
+                .unwrap()
+                .schedule(&complete_graph)
+                .unwrap();
+
+        // TODO: directly use the complete graph when building so that we can generalize the
+        // processing logic for `Chain`s.
+        let fragment_graph = complete_graph.into_inner();
```

Comment:
In the next PRs, we'll hopefully directly use the `CompleteGraph` for building, so that the new dispatchers for the upstream `Materialize` can also be derived naturally based on their distribution.

Reply:
IIUC, current datalog derivation will produce distribution including parallelism for each fragment. In order to build an actor graph, we need parallelism and dispatcher type, so in the future we plan to derive the dispatcher type by fragment distribution for all actors. Am I right?

Reply:
The dispatcher type should be...
- decided by the fragmenter on the frontend and passed in the `Edge` between internal fragments
- hard-coded by us in `CompleteGraph::new` for external fragments, like the `NoShuffle` between upstream `Materialize` and current `Chain`.

With the distributions of up-down fragment pair and the pre-defined dispatcher type, we should be able to fill the `downstream_actors` and `(actor_)hash_mapping` correctly, so that the old scheduler in a post-work manner can be substituted.

Reply:
Got it.